crates/ruma-common/src/events/message/content_serde.rs

```diff
@@ -59,12 +59,42 @@ impl Serialize for MessageContent {
         S: serde::Serializer,
     {
         let mut st = serializer.serialize_struct("MessageContent", 1)?;
-        let variants = self.variants();
-        if variants.len() == 1 && variants[0].mimetype == "text/plain" {
-            st.serialize_field("org.matrix.msc1767.text", &variants[0].body)?;
+        if self.len() == 1 && self[0].mimetype == "text/plain" {
+            st.serialize_field("org.matrix.msc1767.text", &self[0].body)?;
         } else {
-            st.serialize_field("org.matrix.msc1767.message", variants)?;
+            st.serialize_field("org.matrix.msc1767.message", &self.0)?;
         }
         st.end()
     }
 }
+
+pub(crate) mod as_vec {
+    use serde::{ser::SerializeSeq, Deserialize, Deserializer, Serializer};
+
+    use crate::events::message::{MessageContent, Text};
+
+    /// Serializes a `Option<MessageContent>` as a `Vec<Text>`.
+    pub fn serialize<S>(content: &Option<MessageContent>, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        if let Some(content) = content {
+            let mut seq = serializer.serialize_seq(Some(content.len()))?;
+            for e in content.iter() {
+                seq.serialize_element(e)?;
+            }
+            seq.end()
+        } else {
+            serializer.serialize_none()
```

Comment:
I think it would be cleaner to always use `serialize_seq`, with a length of 0 in the `None` case.